src/primary_names.lua

```diff
@@ -0,0 +1,301 @@
+local arns = require("arns")
+local balances = require("balances")
+local utils = require("utils")
+local primaryNames = {}
+
+local PRIMARY_NAME_COST = 100000000 -- 100 IO
+
+---@alias WalletAddress string
+---@alias ArNSName string
+
+---@class PrimaryNames
+---@field owners table<WalletAddress, PrimaryName> - map indexed by owner address containing the primary name and all metadata, used for reverse lookups
+---@field names table<ArNSName, WalletAddress> - map indexed by primary name containing the owner address, used for reverse lookups
+---@field requests table<WalletAddress, PrimaryNameRequest> - map indexed by owner address containing the request, used for pruning expired requests
+
+PrimaryNames = PrimaryNames or {
+	requests = {},
+	names = {},
+	owners = {},
+}
+
+---@class PrimaryName
+---@field name ArNSName
+---@field baseName ArNSName
+---@field startTimestamp number
+
+---@class PrimaryNameWithOwner
+---@field name ArNSName
+---@field baseName ArNSName
+---@field owner WalletAddress
+---@field startTimestamp number
+
+---@class PrimaryNameRequest
+---@field name ArNSName -- the name being requested
+---@field baseName ArNSName -- the base name, identified when creating the name request
+---@field initiator WalletAddress -- the process id that made the request
+---@field startTimestamp number -- the timestamp of the request
+---@field endTimestamp number -- the timestamp of the request expiration
+
+--- Creates a transient request for a primary name. This is done by a user and must be approved by the name owner of the base name.
+--- @param name string -- the name being requested, this could be an undername provided by the ant
+--- @param initiator string -- the address that is creating the primary name request, e.g. the ANT process id
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameRequest|PrimaryNameWithOwner primaryNameRequestOrWithOwner - the request created, or the primary name with owner data if the request is approved
+function primaryNames.createPrimaryNameRequest(name, initiator, timestamp)
+	local baseName = name:match("[^_]+$") or name
+
+	--- existing request for primary name from wallet?
+	local existingRequest = primaryNames.getPrimaryNameRequest(name)
+	assert(not existingRequest, "Primary name request for '" .. name .. "' already exists") -- TODO: should we error here or just let them create a new request and pay the fee again?
+
+	--- check the primary name is not already owned
+	local primaryNameOwner = primaryNames.getAddressForPrimaryName(name)
+	assert(not primaryNameOwner, "Primary name is already owned")
+
+	local record = arns.getRecord(baseName)
+	assert(record, "ArNS record '" .. baseName .. "' does not exist")
+
+	--- TODO: replace with funding plan
+	assert(balances.walletHasSufficientBalance(initiator, PRIMARY_NAME_COST), "Insufficient balance")
+
+	--- transfer the primary name cost from the initiator to the treasury
```

Comment:
Nit: we don't refer to the protocol balance as the treasury anywhere else.

Reply:
[24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)